instrumentation/undertow/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/api/KeyHolder.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.api;
+
+import io.opentelemetry.context.Context;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.AttachmentKey;
+import java.util.IdentityHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Undertow's {@link HttpServerExchange} uses {@link AttachmentKey} as a key for storing arbitrary
+ * data. It uses {@link IdentityHashMap} and thus all keys are compared for equality by object
+ * reference. This means that we cannot hold an instance of {@link AttachmentKey} in a static field
+ * of the corresponding Tracer, as we usually do. Tracers are loaded into user's classloaders and
+ * thus it is totally possible to have several instances of tracers. Each of those instances will
+ * have a separate value in a static field and {@link HttpServerExchange} will treat them as
+ * different keys then.
+ *
+ * <p>That is why this class exists and resides in a separate package. This package is treated in a
+ * special way and is always loaded by bootstrap classloader. This makes sure that this class is
+ * available to all tracers from every classloader.
+ *
+ * <p>But at the same time, being loaded by bootstrap classloader, this class itself cannot initiate
+ * the loading of {@link AttachmentKey} class. Class has to be loaded by <i>any</i> classloader that
+ * has it, e.g. by the classloader of a Tracer that uses this key holder. After that, <i>all</i>
+ * Tracers, loaded by all classloaders, will be able to use exactly the same sole instance of the
+ * key.
+ */
+public class KeyHolder {
+  public static final AtomicReference<AttachmentKey<Context>> contextKey = new AtomicReference<>();
```

Comment:
two thoughts:
* a bootstrap class can't really reference the AttachmentKey class, the reason this is working is because of erasure, but think better to not hide that, and use Object here in the signature
* I think better to use `Map<ClassLoader, Object>` (where the value is AttachmentKey), in case undertow _itself_ is loaded by two different class loaders, in which case you really do need two different AttachmentKey instances. The key of the map would be `AttachmentKey.class.getClassLoader()`.

Reply:
How about a map from class to object to allow scaling this to any type? `Map<Class<?>, Object>`. And since classes from different classloaders aren't equal (I presume), wouldn't need `ClassLoader` in the key I believe.

Reply:
ya, I like that, it's approximately "contextStore" for a single static field

Reply:
Why do I want two different key instance if Undertow is loaded from two different classloaders? And how is the latter possible?

Important point is that any given request must use the same key all the time it passes through Undertow and WildFly.

Reply:
It's not going to happen in wildfly, since wildfly provides undertow in its root class loader, but it's _possible_, e.g. say you had two tomcat apps that each included undertow library.

And if you have two class loaders each with their own copy of undertow classes, and if you store an `AttachmentKey` from one class loader into the `AtomicReference`, and read it back out from the other class loader and cast it to `AttachmentKey` you will get `ClassCastException`.

Reply:
Now I understand :) Thank you for your patience :)

Reply:
Done